Migrate to scalajs-bundler

julienrf opened the pull request:
Hi,

This PR is a work in progress. It shows how we could let scalajs-bundler resolve the npm dependencies and bundle them with the application.

I only updated the `web-examples`. The `mobile-examples` require some work. I wanted first to get your feedback: are you interested in migrating to scalajs-bundler?

nafg commented:
+100

On Mon, Dec 5, 2016, 10:41 AM Julien Richard-Foy <notifications@github.com>
wrote:

> Hi,
>
> This PR is a work in progress. It shows how we could let scalajs-bundler
> resolve the npm dependencies and bundle them with the application.
>
> I only updated the web-examples. The mobile-examples require some work. I
> wanted first to get your feedback: are you interested in migrating to
> scalajs-bundler?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/chandu0101/sri/pull/53
> Commit Summary
>
>    - Migrate to scalajs-bundler
>
> File Changes
>
>    - *M* addons/src/main/scala/sri/addons/perf/ReactPerf.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-0> (2)
>    - *M* build.sbt
>    <https://github.com/chandu0101/sri/pull/53/files#diff-1> (23)
>    - *M* core/src/main/scala/sri/core/React.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-2> (7)
>    - *M* project/CommonUtils.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-3> (5)
>    - *M* project/Dependencies.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-4> (35)
>    - *M* project/LauncherConfigs.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-5> (11)
>    - *D* project/PhantomJS2Env.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-6> (19)
>    - *M* project/plugins.sbt
>    <https://github.com/chandu0101/sri/pull/53/files#diff-7> (2)
>    - *M* test/src/main/scala/sri/test/ReactTestUtils.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-8> (5)
>    - *D* test/src/test/index.js
>    <https://github.com/chandu0101/sri/pull/53/files#diff-9> (14)
>    - *D* test/src/test/package.json
>    <https://github.com/chandu0101/sri/pull/53/files#diff-10> (22)
>    - *M* test/src/test/scala/sri/test/HtmlTagsTest.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-11> (12)
>    - *D* test/src/test/webpack.config.js
>    <https://github.com/chandu0101/sri/pull/53/files#diff-12> (44)
>    - *M* universal/src/main/scala/sri/universal/ReactUniversal.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-13> (4)
>    - *M*
>    universal/src/main/scala/sri/universal/components/DocumentSelectionState.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-14> (2)
>    - *M*
>    universal/src/main/scala/sri/universal/components/ListViewDataSource.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-15> (6)
>    - *M*
>    universal/src/main/scala/sri/universal/components/SwipeableListViewDataSource.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-16> (5)
>    - *M* web-examples/README.md
>    <https://github.com/chandu0101/sri/pull/53/files#diff-17> (9)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/AppRouter.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-18> (0)
>    - *R* web-examples/client/src/main/scala/sri/web/examples/Button.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-19> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/HelloWeb.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-20> (0)
>    - *R* web-examples/client/src/main/scala/sri/web/examples/WebApp.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-21> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/DynamicQueryScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-22> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/DynamicScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-23> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/DynamicStateScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-24> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/HomeScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-25> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/ItemDetailsScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-26> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/ItemsLeftNav.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-27> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/ItemsRouteModule.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-28> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/ItemsScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-29> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/StaticQueryScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-30> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/StaticStateScreen.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-31> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/TopNav.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-32> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/routerexample/WebRouterExample.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-33> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/styles/Colors.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-34> (0)
>    - *R*
>    web-examples/client/src/main/scala/sri/web/examples/styles/Theme.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-35> (0)
>    - *D* web-examples/index.html
>    <https://github.com/chandu0101/sri/pull/53/files#diff-36> (53)
>    - *D* web-examples/index.js
>    <https://github.com/chandu0101/sri/pull/53/files#diff-37> (17)
>    - *D* web-examples/package.json
>    <https://github.com/chandu0101/sri/pull/53/files#diff-38> (34)
>    - *D* web-examples/server.js
>    <https://github.com/chandu0101/sri/pull/53/files#diff-39> (45)
>    - *A* web-examples/server/src/main/resources/application.conf
>    <https://github.com/chandu0101/sri/pull/53/files#diff-40> (1)
>    - *A* web-examples/server/src/main/scala/sri/web/examples/Main.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-41> (72)
>    - *D* web-examples/webpack.config.js
>    <https://github.com/chandu0101/sri/pull/53/files#diff-42> (52)
>    - *D* web-examples/webpack.config.prod.js
>    <https://github.com/chandu0101/sri/pull/53/files#diff-43> (27)
>    - *D* web-examples/webpack.config_backup.js
>    <https://github.com/chandu0101/sri/pull/53/files#diff-44> (46)
>    - *M* web/src/main/scala/sri/web/ReactDOM.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-45> (6)
>    - *D* web/src/main/scala/sri/web/ReactNativeWeb.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-46> (16)
>    - *M* web/src/main/scala/sri/web/router/History.scala
>    <https://github.com/chandu0101/sri/pull/53/files#diff-47> (3)
>
> Patch Links:
>
>    - https://github.com/chandu0101/sri/pull/53.patch
>    - https://github.com/chandu0101/sri/pull/53.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/chandu0101/sri/pull/53>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGAUKTHFK3UJ0vIGDXxsJA4qNr5hQCtks5rFDChgaJpZM4LEX0C>
> .
>

chandu0101 commented:
@julienrf 

>are you interested in migrating to scalajs-bundler?

first of all thanks for PR , and yeah if it works, i am happy to merge :)

>web-examples-server

any specific reason to remove webpack-dev-server ?  if nothing critical can you please revert back to webpack-dev-server ...

julienrf commented:
> any specific reason to remove webpack-dev-server ?

My motivation was to use only one build tool. We can probably define a Scala.js app that uses webpack-dev-server but I thought it would be simpler to embed a basic Play server.

nafg commented:
You can use https://github.com/lihaoyi/workbench

On Tue, Dec 6, 2016, 3:58 AM Julien Richard-Foy <notifications@github.com>
wrote:

> any specific reason to remove webpack-dev-server ?
>
> My motivation was to use only one build tool. We can probably define a
> Scala.js app that uses webpack-dev-server but I thought it would be simpler
> to embed a basic Play server.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/chandu0101/sri/pull/53#issuecomment-265095629>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGAUFvHoEWDeGr14CeUYHp6MB-XPHK0ks5rFSOagaJpZM4LEX0C>
> .
>

chandu0101 commented:
>My motivation was to use only one build tool. We can probably define a Scala.js app that uses webpack-dev-server but I thought it would be simpler to embed a basic Play server.

ok , may be we should use https://github.com/lihaoyi/workbench !

julienrf commented:
Not sure workbench will have the same “historyApiFallback” feature as webpack-dev-server.

chandu0101 commented:
good point, just keep webpack-dev-server  for now if possible.

julienrf commented:
@chandu0101 I would like to highlight the fact that if your facade uses scalajs-bundler, then your users will also have to use this plugin (or manually define an npm build that fetches the dependencies and bundles them with the app). I’m afraid that the transition might be annoying for some users. Maybe we should keep the current version and still merge this PR into a distinct branch, and release it under a different version number, to see how it is received by users. What do you think?

I will revert to webpack-dev-server, but it means that I will have an npm build in addition to the sbt build. Is this really what you want?

chandu0101 commented:
> I’m afraid that the transition might be annoying for some users. Maybe we should keep the current version and still merge this PR into a distinct branch, and release it under a different version number, to see how it is received by users. What do you think?

👍  , lets do that :) 

>I will revert to webpack-dev-server, but it means that I will have an npm build in addition to the sbt build. Is this really what you want?

to be honest i didn't fully understand this . why we need npm build ? scalajs-bundler generates a single output file right ? can't we directly use that in index.html file ? :s

julienrf commented:
> can't we directly use that in index.html file ?

No because the router of the application uses URL that would make the browser try to load resources that do not exist.

julienrf commented:
(maybe we could just switch to a fragment-based router)

chandu0101 commented:
>(maybe we could just switch to a hash-based router)

ok , lets change this line https://github.com/chandu0101/sri/blob/master/web-examples/src/main/scala/sri/web/examples/AppRouter.scala#L19  

to 
```scala

  override val history = HistoryFactory.hashHistory()

```

btw i created new branch `scalajs-bundler`

julienrf commented:
OK, I updated the PR to simplify the web-examples project and changed the target branch to `scalajs-bundler`.

I think it would be great to check that the mobile examples still work. What are the setup steps to follow to build them?

chandu0101 commented:
https://github.com/chandu0101/sri/tree/master/mobile-examples#mobile-examples

chandu0101 commented:
let me know if you need any help in testing mobile-examples ...

chandu0101 commented on project/Dependencies.scala:
```diff
@@ -42,11 +45,26 @@ object Dependencies {
+  val universalModuleDeps = Seq(
+    npmDependencies in Compile ++= Seq(
+      "react" -> "15.4.0-rc.4",
+      "react-native" -> "0.38.0"
```
hmm i think we have a problem here , universal module contains common components between mobile and web module , web version is backed by `react-native-web` . can you please move `react-native` dependency to mobile module ? and also can you please change scope of npmDependencies to provided ...

chandu0101 replied:
how about moving `npmDependencies` to examples projects!

julienrf replied:
Moving the dependencies definition to the examples projects will indeed have the effect of setting them as “provided”

chandu0101 replied:
indeed, i am just wondering do we really need this in core modules :s

julienrf replied:
Why don’t you want to define these dependencies at the facade level? This would save users to do this step on their projects.

chandu0101 replied:
if we specify in module level , does user can override dependency with different version ?

julienrf replied:
Currently not (but there is a ticket for this https://github.com/scalacenter/scalajs-bundler/issues/55).

chandu0101 replied:
in that case its ok. In past scalajs-jquery facade shipped with jsDeps(jQuery), soon users of library faced issues ..  , thats why i asked :)

chandu0101 commented on project/LauncherConfigs.scala:
```diff
@@ -61,28 +57,6 @@ object LauncherConfigs {
     )
 
-  /**
-   * react-native prod bundler needs require function without name spaces
```
we need those process functions. react-native asset system uses `require`

example : 
```scala
    @inline def load[T](lib: String): T = g.require(lib).asInstanceOf[T]

    @inline def loadDynamic(lib: String): js.Dynamic = load[js.Dynamic](lib)

loadDynamic("./images/thumbnails/dislike.png"),

```

in production mode we need to remove prefix etc

julienrf replied:
If I understand correctly they are not needed anymore because Scala.js emits a proper CommonJS module (so there is no `exportsNamespace` prefix anymore).

chandu0101 replied:
cool ok , i should have tested before asking :( , thank you :)

dispalt commented:
It looks like you merged most of this? Right?

chandu0101 commented:
yes, but i didn't added scalajs-bundler to the library, i think now user can choose to use plain npm/webpack or scalajs-bundler or some other build/package tools.

julienrf commented:
Feel free to close this PR if it is redundant with what you’ve pushed.

chandu0101 commented:
@julienrf thank you :)
